src/compiler.ts

```diff
@@ -5537,8 +5537,17 @@ export class Compiler extends DiagnosticEmitter {
     valueExpr: ExpressionRef,
     reportNode: Node
   ): ExpressionRef {
-    // FIXME: see comment in compileBinaryOverload below why recompiling on type mismatch
-    // is a bad idea currently. so this assumes that the type matches.
+    var numParams = operatorInstance.signature.parameterTypes.length;
+    var expectedNumParams = operatorInstance.is(CommonFlags.INSTANCE) ? 0 : 1;
+    if (numParams != expectedNumParams) {
+      this.error(
+        DiagnosticCode.Expected_0_arguments_but_got_1,
+        operatorInstance.parametersOrIdentifierRange,
+        expectedNumParams.toString(),
+        numParams.toString()
+      );
+      return this.module.unreachable();
```

Comment:
Looks like `this.currentType` should still be set to the return type to prevent issues further down.